sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/impl/rule/JoinReorderingTest.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.extensions.sql.impl.rule;
+
+import java.math.BigInteger;
+import org.apache.beam.sdk.extensions.sql.impl.BeamSqlEnv;
+import org.apache.beam.sdk.extensions.sql.impl.rel.BeamIOSourceRel;
+import org.apache.beam.sdk.extensions.sql.impl.rel.BeamRelNode;
+import org.apache.beam.sdk.extensions.sql.meta.provider.test.TestTableProvider;
+import org.apache.beam.sdk.values.Row;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Join;
+import org.junit.Assert;
+import org.junit.Test;
+
+/**
+ * This test ensures that we are reordering joins and get a plan similar to Join(large,Join(small,
```

Comment:
I think the critical part is to only include the rules you need as Rui suggested (re-ordering rule in this case), otherwise you don't know for sure why the large join is at the top. And we should also include the opposite tests, that without the rule the order is the same as in original query.

Reply:
+1

Reply:
I added two more types of tests. One is testing if each of the rules alone acts correctly (by once converting a query without that and once converting it with the rule) and the other tests if the entire system without reordering rules does not reorder joins.

However, I still do not think it is a good idea to test the entire system without the join reordering rules to check if they are not producing optimal join order. (testSystemNotReorderingWithoutRules())

We expect the system to reorder the joins. But I am not sure why it should be an expected behavior of the system not to reorder joins without these specific rules. (We are not checking this for many other rules and features in the system. For instance we are not checking if the filter will be pushed down of the join in case of removal of FilterJoinRule. And I think this can also be applied to many "subsets of rules" that do some sort of behavior when applied together.) 

Furthermore, this test will break if someone adds a rule that reorder some certain types of joins or reorder all joins in some other way and I am not sure why we should prevent them from doing that as long as the system is still producing our expected optimized order.

Reply:
It makes sense. The behavior of the system is our design decision. One way to document our design is to write tests. Until now there were no tests around these things and I don't think we had a clear idea of what's happening and didn't have such contracts defined around rules. Now that we do it's fine to go one way or another if it's conscious and explicit. In this case the actual behavior is not going to change no matter how we decide to approach the "rules integration" part and I doubt that anytime soon we will face these kinds of issues. And if we do, we can re-assess the decisions at that point if needed. I don't have good reasons to lean towards one way or another at this point.

Reply:
Agree with Anton that changes and designs are documented by tests. As we are at an early stage of having optimizations, what we have now is definitely not perfect and we can refine them as time goes.